File: scripts/seed-ahl.mjs
#!/usr/bin/env node
/**
 * scripts/seed-ahl.mjs
 * Seeds AHL game data from HockeyTech API
 * 
 * API: https://lscluster.hockeytech.com/feed/index.php
 * Params: feed=statviewfeed&view=schedule&season=90&team=-1&...
 */

import { createClient } from '@supabase/supabase-js';
import { randomUUID } from 'crypto';

const SUPABASE_URL = 'https://yszheonqyyskkjoxoexk.supabase.co';
const SERVICE_KEY  = '***REMOVED***';
const AHL_LEAGUE_ID = 'b05d6d26-d5d6-4cfd-a48b-f5646fa7d611';

const supabase = createClient(SUPABASE_URL, SERVICE_KEY);

// AHL team city → Supabase team name
const CITY_TO_TEAM = {
  'Abbotsford':       'Abbotsford Canucks',
  'Bakersfield':      'Bakersfield Condors',
  'Belleville':      'Belleville Senators',
  'Bridgeport':      'Bridgeport Islanders',
  'Calgary':         'Stockton Heat',           // displaced AHL team
  'Charlotte':       'Charlotte Checkers',
  'Chicago':         'Chicago Wolves',
  'Cleveland':      'Cleveland Monsters',
  'Coachella Valley':'Coachella Valley Firebirds',
  'Colorado':        'Colorado Eagles',
  'Grand Rapids':   'Grand Rapids Griffins',
  'Hartford':       'Hartford Wolf Pack',
  'Hershey':        'Hershey Bears',
  'Iowa':           'Iowa Wild',
  'Laval':          'Laval Rocket',
  'Lehigh Valley':  'Lehigh Valley Phantoms',
  'Manitoba':       'Manitoba Moose',
  'Milwaukee':      'Milwaukee Admirals',
  'Ontario':        'Ontario Reign',
  'Providence':     'Providence Bruins',
  'Rochester':      'Rochester Americans',
  'San Diego':      'Stockton Heat',           // displaced team
  'San Jose':       'San Jose Barracuda',
  'Santa Clara':    'Santa Clara Roadrunners',
  'Springfield':    'Springfield Thunderbirds',
  'Syracuse':       'Syracuse Crunch',
  'Texas':          'Texas Stars',
  'Toronto':        'Toronto Marlies',
  'Tucson':         'Tucson Roadrunners',
  'Utica':          'Utica Comets',
  'Wilkes-Barre/Scranton': 'Wilkes-Barre/Scranton Penguins',
  'Henderson':      'Ontario Reign',            // Henderson Silver Knights → Ontario Reign
  'Rockford':       'Illinois Rockets',         // Rockford IceHogs renamed in our DB
};

// Season ID for AHL 2024-25
const SEASON = '90';

async function getTeamIdMap() {
  const { data } = await supabase.from('teams').select('id, name, city').eq('league_id', AHL_LEAGUE_ID);
  const map = {};
  // Key by city
  if (data) {
    for (const t of data) {
      const c = (t.city || '').split('(')[0].trim().toLowerCase();
      map[c] = t.id;
    }
  }
  return map;
}

function parseDate(dateWithDay) {
  // e.g. "Thu, Mar 19" → assume 2025
  const months = { Jan:0, Feb:1, Mar:2, Apr:3, May:4, Jun:5, Jul:6, Aug:7, Sep:8, Oct:9, Nov:10, Dec:11 };
  const parts = dateWithDay.match(/(\w+),\s+(\w+)\s+(\d+)/);
  if (!parts) return null;
  const mon = months[parts[2]] ?? 2; // default Mar
  const day = parseInt(parts[3]);
  // Season is 2024-2025 (Oct 2024 – Apr 2025) or 2025-2026 (Oct 2025 – Apr 2026)
  let year = mon >= 9 ? 2024 : 2025;
  if (mon >= 9 && day >= 1) year = 2024;
  else if (mon <= 4 && day <= 31) year = 2025;
  return new Date(year, mon, day, 19, 0, 0);
}

async function fetchAHLGames(month) {
  const url = `https://lscluster.hockeytech.com/feed/index.php?feed=statviewfeed&view=schedule&season=${SEASON}&team=-1&client_code=ahl&league_id=4&site_id=3&key=ccb91f29d6744675&location=homeaway&date=&month=${month}`;
  
  const res = await fetch(url, {
    headers: { 'User-Agent': 'Mozilla/5.0', 'Accept': 'application/json' },
    signal: AbortSignal.timeout(15000)
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  
  const text = await res.text();
  
  // Parse the JSONP-like response — extract JSON from callback
  // Format: somecallback({ ... })
  let json;
  const m = text.match(/somecallback\s*\(\s*(\{.*})\s*\)\s*;?\s*$/s);
  if (m) {
    json = JSON.parse(m[1]);
  } else {
    // Try direct parse
    json = JSON.parse(text);
  }
  
  return json;
}

function extractGamesFromResponse(data) {
  if (!data || !data sections) return [];
  
  for (const section of data.sections || []) {
    if (!section.data) continue;
    for (const row of section.data) {
      // games found here
      return section.data.map(r => ({
        gameId:    r.row?.game_id,
        dateStr:   r.row?.date_with_day,
        homeCity:  r.row?.home_team_city,
        awayCity:  r.row?.visiting_team_city,
        homeScore: r.row?.home_goal_count,
        awayScore: r.row?.visiting_goal_count,
        status:    r.row?.game_status,
        homeLink:  r.prop?.home_team_city?.teamLink,
        awayLink:  r.prop?.visiting_team_city?.teamLink,
      })).filter(g => g.gameId);
    }
  }
  return [];
}

// Alternate extraction: the response is an array of section objects
function extractGamesDirect(data) {
  if (!data) return [];
  const games = [];
  
  function walk(obj) {
    if (!obj || typeof obj !== 'object') return;
    if (Array.isArray(obj)) {
      obj.forEach(walk);
      return;
    }
    if (obj.row && obj.row.game_id && obj.row.date_with_day) {
      games.push({
        gameId:    obj.row.game_id,
        dateStr:   obj.row.date_with_day,
        homeCity:  obj.row.home_team_city,
        awayCity:  obj.row.visiting_team_city,
        homeScore: obj.row.home_goal_count,
        awayScore: obj.row.visiting_goal_count,
        status:    obj.row.game_status,
      });
    }
    Object.values(obj).forEach(walk);
  }
  
  walk(data);
  return games;
}

async function main() {
  const teamIdMap = await getTeamIdMap();
  console.log(`Loaded ${Object.keys(teamIdMap).length} AHL teams`);
  
  // Fetch months Sep 2024 through May 2025 (full season)
  const allGames = [];
  
  for (let m = 9; m <= 17; m++) {
    const month = m <= 12 ? m : m - 12;
    const year  = m <= 12 ? 2024 : 2025;
    if (month > 5) break; // May max
    
    const monthStr = `${year}-${String(month).padStart(2,'0')}`;
    console.log(`Fetching AHL data for month ${monthStr}...`);
    
    try {
      const raw = await fetchAHLGames(monthStr);
      const games = extractGamesDirect(raw);
      console.log(`  Found ${games.length} games`);
      allGames.push(...games);
      await new Promise(r => setTimeout(r, 300));
    } catch (e) {
      console.error(`  Error fetching month ${monthStr}: ${e.message}`);
    }
  }
  
  console.log(`Total games fetched: ${allGames.length}`);
  
  let inserted = 0;
  for (const g of allGames) {
    if (!g.gameId) continue;
    
    const homeCityKey = (g.homeCity || '').toLowerCase().replace(/\/$/, '');
    const awayCityKey = (g.awayCity || '').toLowerCase().replace(/\/$/, '');
    
    const homeTeamId = teamIdMap[homeCityKey] || null;
    const awayTeamId = teamIdMap[awayCityKey] || null;
    
    const scheduledAt = parseDate(g.dateStr);
    if (!scheduledAt) continue;
    
    let status = 'completed';
    if (g.status && (g.status.includes('OT') || g.status.includes('SO'))) {
      status = 'completed'; // OT/SO is still completed
    } else if (g.status && g.status !== 'Final') {
      status = 'scheduled'; // future game
    }
    
    const fixture = {
      id:           randomUUID(),
      home_team_id: homeTeamId,
      away_team_id: awayTeamId,
      league_id:    AHL_LEAGUE_ID,
      venue_id:     null,
      scheduled_at: scheduledAt.toISOString(),
      home_score:   g.homeScore != null ? parseInt(g.homeScore) : null,
      away_score:   g.awayScore != null ? parseInt(g.awayScore) : null,
      status,
      season:       '2024-25',
      game_data:    JSON.stringify({
        ahl_game_id: String(g.gameId),
        status:      g.status,
        overtime:   !!(g.status && (g.status.includes('OT') || g.status.includes('SO'))),
        home_city:  g.homeCity,
        away_city:  g.awayCity,
        home_team_link: g.homeLink,
        away_team_link: g.awayLink,
      }),
    };
    
    const { error } = await supabase.from('fixtures').upsert(fixture, { onConflict: 'id' });
    if (!error) inserted++;
  }
  
  console.log(`Inserted/updated ${inserted} AHL fixtures`);
}

main().catch(console.error);

File: scripts/seed-ahl.py
#!/usr/bin/env python3
"""
scripts/seed-ahl.py
Seeds AHL game data from HockeyTech API into Supabase.
"""

import load_secrets
import os, urllib.request, json, time, datetime, uuid, re

SUPABASE_URL  = os.environ['NEXT_PUBLIC_SUPABASE_URL']
ANON_KEY     = os.environ['NEXT_PUBLIC_SUPABASE_ANON_KEY']
AHL_LEAGUE_ID = 'b05d6d26-d5d6-4cfd-a48b-f5646fa7d611'
SEASON       = '90'

# HockeyTech team city name (lowercase) → Supabase team city key (lowercase)
HT_CITY_MAP = {
    'abbotsford':             'abbotsford',
    'bakersfield':            'bakersfield',
    'belleville':             'belleville',
    'bridgeport':              'bridgeport',
    'calgary':               'stockton',         # displaced team
    'charlotte':              'charlotte',
    'chicago':                'rosemont',          # Chicago Wolves (Rosemont, IL)
    'cleveland':              'cleveland',
    'coachella valley':       'palm desert',
    'colorado':               'loveland',
    'grand rapids':           'grand rapids',
    'hartford':               'hartford',
    'hershey':                'hershey',
    'iowa':                   'des moines',
    'laval':                  'laval',
    'lehigh valley':          'allentown',
    'manitoba':               'winnipeg',
    'milwaukee':              'milwaukee',
    'ontario':                'ontario',
    'providence':              'providence',
    'rochester':              'rochester',
    'san diego':              'stockton',         # displaced team
    'san jose':               'san jose',
    'santa clara':            'santa clara',
    'springfield':            'springfield',
    'syracuse':               'syracuse',
    'texas':                  'cedar park',
    'toronto':                'toronto (coca-cola coliseum)',
    'tucson':                 'tucson',
    'utica':                  'utica',
    'wilkes-barre/scranton':  'wilkes-barre',
    'rockford':               'rosemont',          # Rockford IceHogs
    'henderson':              'ontario',            # Henderson Silver Knights
};


def get_team_map():
    headers = {'apikey': ANON_KEY, 'Authorization': f'Bearer {ANON_KEY}'}
    req = urllib.request.Request(
        f'{SUPABASE_URL}/rest/v1/teams?select=id,name,city&league_id=eq.{AHL_LEAGUE_ID}',
        headers=headers)
    resp = urllib.request.urlopen(req, timeout=10)
    teams = json.loads(resp.read())
    m = {}
    for t in teams:
        c = (t['city'] or '').strip().lower()
        m[c] = t['id']
        simple = c.split('(')[0].strip()
        m[simple] = t['id']
    return m


def parse_date(date_str):
    """Parse 'Thu, Mar 19' → datetime."""
    months = {'Jan':1,'Feb':2,'Mar':3,'Apr':4,'May':5,'Jun':6,
              'Jul':7,'Aug':8,'Sep':9,'Oct':10,'Nov':11,'Dec':12}
    m = re.search(r'(\w+),\s+(\w+)\s+(\d+)', date_str)
    if not m:
        return None
    mon = months.get(m.group(2))
    day = int(m.group(3))
    if mon is None:
        return None
    year = 2024 if mon >= 10 else 2025
    return datetime.datetime(year, mon, day, 19, 0, 0)


def fetch_month(month_num):
    """Fetch AHL schedule for month (1-12)."""
    ms = f'{month_num:02d}'
    url = (
        f'https://lscluster.hockeytech.com/feed/index.php'
        f'?feed=statviewfeed&view=schedule&season={SEASON}&team=-1'
        f'&client_code=ahl&league_id=4&site_id=3'
        f'&key=ccb91f29d6744675&location=homeaway&date=&month={ms}'
    )
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    resp = urllib.request.urlopen(req, timeout=15)
    raw = resp.read().decode('utf-8')

    # Response: ([{...}]) — strip outer parens
    stripped = raw.strip()
    if stripped.startswith('(['):
        stripped = stripped[1:]
    if stripped.endswith('])'):
        stripped = stripped[:-1]

    data = json.loads(stripped)
    games = []
    for section in data[0].get('sections', []):
        for row in section.get('data', []):
            r = row.get('row', {})
            if not r.get('game_id'):
                continue
            games.append({
                'game_id':      r['game_id'],
                'date':         r.get('date_with_day'),
                'home':         r.get('home_team_city'),
                'away':         r.get('visiting_team_city'),
                'home_score':   r.get('home_goal_count'),
                'away_score':   r.get('visiting_goal_count'),
                'status':       r.get('game_status'),
            })
    return games


def upsert_fixture(fixture):
    """Insert one fixture via Supabase REST."""
    headers = {
        'apikey': ANON_KEY,
        'Authorization': f'Bearer {ANON_KEY}',
        'Content-Type': 'application/json',
        'Prefer': 'return=representation',
    }
    req = urllib.request.Request(
        f'{SUPABASE_URL}/rest/v1/fixtures',
        data=json.dumps(fixture).encode(),
        headers=headers,
        method='POST'
    )
    try:
        resp = urllib.request.urlopen(req, timeout=10)
        return resp.status in (200, 201)
    except urllib.error.HTTPError as e:
        if e.code == 409:
            return None  # already exists, skip
        print(f'  HTTP {e.code}: {e.read()[:200]}')
        return False


def main():
    print('Loading AHL team IDs...')
    team_map = get_team_map()
    print(f'  {len(team_map)} city→team mappings loaded')

    # Build list of (month_num, year) pairs for Oct 2024 – Apr 2025
    months = []
    d = datetime.date(2024, 10, 1)
    end = datetime.date(2025, 4, 30)
    while d <= end:
        months.append((d.month, d.year))
        d = (d.replace(day=28) + datetime.timedelta(days=32)).replace(day=1)

    total_inserted = 0
    total_skipped  = 0
    total_errors   = 0

    for month_num, year in months:
        month_label = f'{year}-{month_num:02d}'
        print(f'\n── {month_label} ──')
        try:
            games = fetch_month(month_num)
            print(f'  {len(games)} games fetched')
        except Exception as e:
            print(f'  Fetch error: {e}')
            continue

        for g in games:
            home_key = HT_CITY_MAP.get((g['home'] or '').lower().strip())
            away_key = HT_CITY_MAP.get((g['away'] or '').lower().strip())

            home_team_id = team_map.get(home_key) if home_key else None
            away_team_id = team_map.get(away_key) if away_key else None

            scheduled = parse_date(g['date'] or '')
            if not scheduled:
                continue

            status_raw = (g['status'] or '').lower()
            status = 'completed' if 'final' in status_raw else 'scheduled'

            fixture = {
                'id':            uuid.uuid4().hex,
                'home_team_id':  home_team_id,
                'away_team_id':  away_team_id,
                'league_id':     AHL_LEAGUE_ID,
                'venue_id':      None,
                'scheduled_at':  scheduled.isoformat(),
                'home_score':    int(g['home_score']) if g['home_score'] else None,
                'away_score':    int(g['away_score']) if g['away_score'] else None,
                'status':        status,
                'season':        '2024-25',
                'game_data': json.dumps({
                    'ahl_game_id': str(g['game_id']),
                    'status':      g['status'],
                    'overtime':    'ot' in status_raw or 'so' in status_raw,
                }),
            }

            result = upsert_fixture(fixture)
            if result is True:
                total_inserted += 1
            elif result is None:
                total_skipped += 1
            else:
                total_errors += 1

        time.sleep(0.3)

    print(f'''
════════════════════════════════════
  AHL seeding complete
  Inserted : {total_inserted}
  Skipped   : {total_skipped}
  Errors    : {total_errors}
════════════════════════════════════''')


if __name__ == '__main__':
    main()